src/funnel.ts

```diff
@@ -0,0 +1,280 @@
+
+  /**
+   * Resets the funnel to it's initial state. Any calls made since the last
+   * invocation will be discarded.
+   */
+  readonly cancel: () => void;
+
+  /**
+   * Triggers an invocation regardless of the current state of the funnel.
+   * Like any other invocation, The funnel will also be reset to it's initial
+   * state afterwards.
+   */
+  readonly flush: () => void;
+
+  /**
+   * The funnel is in it's initial state (there are no active timeouts).
+   */
+  readonly isIdle: boolean;
+};
+
+/**
+ * Creates a funnel that controls the timing and execution of `callback`. Its
+ * main purpose is to manage multiple consecutive (usually fast-paced) calls,
+ * reshaping them according to a defined batching strategy and timing policy.
+ * This is useful when handling uncontrolled call rates, such as DOM events or
+ * network traffic. It can implement strategies like debouncing, throttling,
+ * batching, and more.
+ *
+ * An optional `reducer` function can be provided to allow passing data to the
+ * callback via calls to `call` (otherwise the signature of `call` takes no
+ * arguments).
+ *
+ * Typing is inferred from `callback`s param, and from the rest params that
+ * the optional `reducer` function accepts. Use **explicit** types for these
+ * to ensure that everything _else_ is well-typed.
+ *
+ * Notice that this function constructs a funnel **object**, and does **not**
+ * execute anything when called. The returned object should be used to execute
+ * the funnel via the its `call` method.
+ *
+ * Debouncing: use `burstCoolDownMs` and any `invokedAt`.
+ * Throttling: use `delayMs` and `invokedAt: "start"` or `"both"`.
+ * Batching: See the reference implementation in [`funnel.reference-batch.test.ts`](https://github.com/remeda/remeda/blob/main/src/funnel.reference-batch.test.ts).
+ *
+ * @param callback - The main function that would be invoked occasionally based
+ * on `options`. The function would take the latest result of the `reducer`; if
+ * no calls where made since the last time it was invoked it will not be
+ * invoked. If a return value is needed, it should be passed via a reference or
+ * via closure to the outer scope of the funnel.
+ * @param options - An object that defines when `execute` should be
+ * invoked, relative to the calls of `call`. A timer that isn't defined will
+ * **not** be enabled, 0 is not used a default fallback.
+ * @param options.reducer - Combines the arguments passed to `call` with the
+ * value computed on the previous call (or `undefined` on the first time). The
+ * goal of the function is to extract and summarize the data needed for
+ * `callback`. It should be fast and simple as it is called often and should
+ * defer heavy operations to the `execute` function. If the final value
+ * is `undefined`, `callback` will not be called.
+ * @param options.invokedAt - At what "edges" of the funnel's activity
+ * window should `execute` be invoked. `start` means The function will be
```

Comment:
**nit:** can this be formatted as a list, with `start`, and `end` etc.?

Reply:
we should really improve how the docs site presents params...

Reply:
I'm not sure that would render properly in IDE tooltips, they usually consider this as a single line.